### PR TITLE
Adding initial implementation of the password recovery

### PR DESCRIPTION
**Pull request by lmanohara**
- DAO class implementation.
- Notification manager implementation.

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/RecoveryScenarios.java [hunk 1]
```diff
@@ -1,15 +1,30 @@
...
+* limitations under the License.
+*/
+package org.wso2.carbon.identity.recovery;
+
+/**
+ * Recovery Scenarios Constants.
+ */
+public enum RecoveryScenarios {
+    NOTIFICATION_BASED_PW_RECOVERY,
+    QUESTION_BASED_PWD_RECOVERY,
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/RecoveryScenarios.java]: Better to maintain consistency, ie. either let's user PWD == 'Password' or PW == 'Passoword'
IMO let's use PASSWORD without going for PWD or PW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/RecoverySteps.java [hunk 2]
```diff
@@ -1,12 +1,27 @@
...
+package org.wso2.carbon.identity.recovery;
+
+/**
+ * Recovery Steps constants.
+ */
+public enum RecoverySteps {
+    NOTIFY,
+    UPDATE_PASSWORD,
+    VALIDATE_CHALLENGE_QUESTION,
+    VALIDATE_ALL_CHALLENGE_QUESTION,
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/RecoverySteps.java]: VALIDATE_ALL_CHALLENGE_QUESTION --> VALIDATE_ALL_CHALLENGE_QUESTIONS ?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/IdentityRecoveryServiceDataHolder.java [hunk 3]
```diff
@@ -25,19 +27,20 @@
 
     private static IdentityRecoveryServiceDataHolder instance = new IdentityRecoveryServiceDataHolder();
     private RealmService realmService;
-//    private EventService identityEventService;
+    private EventService identityEventService;
+    private EmailLinkConfig emailLinkConfig;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/IdentityRecoveryServiceDataHolder.java]: What does think EmailLinkConfig class do?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/mapping/EmailLinkConfig.java [hunk 4]
```diff
@@ -24,6 +24,9 @@
 public class EmailLinkConfig {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/mapping/EmailLinkConfig.java]: EmailNotificationConfig or EmailNotificationConfig would be better name for this class IMO :wink:

**Comment by lmanohara** [comment]
Fixed https://github.com/wso2/carbon-identity-mgt/pull/124/commits/ba05e84330f0a73ae3e8f9abf56a094537f99192

**Pull request merged**